### PR TITLE
[FIX] l10n_at: Upgrade Austrian Localization

### PR DESCRIPTION
**Pull request by wtaferner**
**Description of the issue/feature this PR addresses:**
Several fixes and updates for the not yet complete Austrian Localization, started to prepare in Valencia on the OCA Code Sprint.

Enterprise PR: https://github.com/odoo/enterprise/pull/47197

Info: @wt-io-it

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by svfu-odoo** [comment]
I have added more information on one of your other PRs. Please see  https://github.com/odoo/enterprise/pull/47198#issuecomment-1719398629

**Comment by wtaferner** [comment]
@JulienVR @svfu-odoo 
Just because 14.0 is not supported after release of 17.0 and I was not sure whether it has any value at this point.
About the other task, please merge my PRs first and redo the automatic transformation in master based on 16.0 again and also be so kind and do not completely steel the blame/copyright by deleting my work and converting/copying it just to a new format and new files.

Please leave me as author as nothing of value was added besides adapting it to some new technical structures and adding a list of SAF-T codes which are actually pretty old where I provided the links to the documentation. 

So yeah, appreciate it that finally someone officially start to work on Austrian legal requirements, but be aware that I have quite some insights, a lot of experience as I am the one who needs to work with it on a daily basis and should kinda have the lead if you can accept something like that.

I am also not convinced about the new tags for the balance, especially there are tags which should be shown either on the Activa or the Passiva side of the balance depending on positive or negative balance. Also there are conditional sections in the balance report. I already told some of you guys that it is a big step back if we introduce new tags and to huge domains on the reports without any flexibility (not the way how we structure the balance and the profit and loss report here). 

I currently see a huge mess in 17.0 ahead which you do hopefully respect, understand and try to work it out with me and hopefully in other localizations as well, otherwise you will not only disappoint me massively but also will make it a nightmare to work in Austria with the localization and I am doing so for years working quite successfully with tax lawyers whereas Odoo accounting is leading and doing the full accounting and only the final statements are done in external accounting software.

Alright, please get connected besides writing or implying changes here on Github to get this thing done together and for the most, please merge this asap...it is in production for all my clients in 15.0 and 16.0.

@william-andre 
Not sure if this way will work out...this will or could hurt me and the Austrian localization really a lot...but I will try to help if respect and dignity is shown to my work and myself.

**Comment by wtaferner** [comment]
@jco-odoo 
Please help, I have seen that there have been done changes in master which are really not what I would expect. Do I really need to take now what was done in huge not reviewable rewrites and fix the Austrian Localization again?
I already said in Valenica that I do understand that some generic and common approaches are useful, but I see huge changes with a lot of rewrites now without letting me review before which will not improve the Austrian Localization, but make it worse and for the most you will loose me as a contributor, after I just tried again to submit my work done behind the curtains the last years by working with Austrian clients in detail.

**Comment by svfu-odoo** [comment]
@wtaferner 

There was no intention to steal the blame.
Due to the translation of the files into English (which is the standard in odoo) lots of lines were changed.

We will merge your PRs first, since they are in production already.

The PRs on master are not set in stone.
We will of course consider your feedback / review on the incoming changes since we respect your work in Austria and your contributions to and insight into the Austrian localization.
We would be happy to discuss the changes with you. I will send you an email to get connected.

**Comment by wtaferner** [comment]
@Levizar 
I tried to take your review into consideration here as well and backported the little fix in the update tax method in an own commit within this PR. This PR should be ok to be merged, after passing the tests. I have also tested the procedure manually again with quite some real data.

@clbr-odoo 
What I recognized about the update mechanism...maybe you can update me on the wanted behavior

- Fiscal positions are not updated
- Tax Scope is not updated in case added or changed

These are the most obvious issues with an impact, even though as always with localization underlying individual settings which might be the reason why you skipped those.

Also I recognized...if you had already a tax with a matching ID (the one we actually update) and the xml id was updateable, it will be deleted in the end of the update procedure...this does not happen if the related external id `l10n_at.1_xxxx_tax_name` is not updateable, so I guess this is unwanted behavior, right?

**Comment by clbr-odoo** [comment]
Hello @wtaferner ,

> Fiscal positions are not updated

Not "updated" but sometimes "created". See [here for  details](https://github.com/odoo/odoo/blob/15.0/addons/account/models/chart_template.py#L159). This is a defensive choice to not mess up existing configuration but we're still open for discussion based on use-cases.

> Tax Scope is not updated in case added or changed

Indeed, it doesn't sound like something that is supposed to happen right ? We could manage the case where it is added and previously non-existent. A tax with a fixed xmlid don't suddenly change scope so the update seems very error-prone.

Can you give me a bit more details about your last point please ? Are your `account.tax` noupdate=False ? This is not how Odoo works by default so it seems very likely that it is an unwanted behavior yes. I don't get what you mean with when "the external id is not updatable"?

**Comment by wtaferner** [comment]
@clbr-odoo 

> Fiscal positions are not updated

Yeah, I remember, maybe @william-andre can remember as well, completely new fiscal positions are not created for no reason...updating seems to be a good choice in terms of defensive but for completely new fiscal positions creation should be a no brainer...

> Indeed, it doesn't sound like something that is supposed to happen right ? We could manage the case where it is added and previously non-existent. A tax with a fixed xmlid don't suddenly change scope so the update seems very error-prone.

There is exactly one use case where I would expect it to be updated...when it was `False` and the update of the module is introducing `service` or `consu` as a value which is the case in this PR.

> noupdate=False

Yes, and I know normally it is created by the module, but I had some manually created xmlids to match the update to preserve and align my manually created taxes representing what will be updated. And this is why I stumbled into this situation...in case they are not locked, they will get deleted after the update/upgrade which is weird IMO.

So, I guess it might be a good idea to "fix" this automatically before update/upgrade that all xmlids related are noupdate=True or during matching whereas I do not know if it is maybe already too late in the stage where the script is executed.

But yeah, nevermind, I know how to resolve it...it was just to safe guard maybe a bit better for such situations.

**Comment by clbr-odoo** [comment]
@wtaferner 

> Fiscal positions are not updated

Oh sorry, are you talking about `account.fiscal.position` ? Indeed only `account.fiscal.position.tax` are created. I'm pretty sure we do it in saas-16.2. The first version of the update was more a "let's start small" task since we didn't took care of our data evolution through time before that.

> Tax Scope is not updated in case added or changed

I agree with this single use-case. :+1: 

> noupdate=False

I don't really have a definitive answer for this one, it is a typical case of "protect me" vs. "don't do things in my back". I imagine there is a reason for those taxes to be in noupdate=False so I can't really tell if it is always ok to override the on-purpose setting :)

**Comment by wtaferner** [comment]
@JulienVR 
Any ETA when this might be merged? I think @svfu-odoo is also waiting for this to move forward.

File: addons/l10n_at/data/account_tax_template.xml
Comment (jco-odoo) [inline addons/l10n_at/data/account_tax_template.xml:2029]: how?

Comment (wtaferner) [reply addons/l10n_at/data/account_tax_template.xml:2029]: Before, it was not properly defined at all and here we have the use case that the carrier is taking over your duty to pay the tax to the ministry of finance in the name of the company, therefore it is forwarded (purely the tax). 
As we are allowed to deduct the whole amount in terms of deductible VAT, it is like that an properly encoded on the related accounts.

Same is true for 083 where we do get an information by the ministry of finance that we have to pay the "sale tax". Based on this we do an extra tax account move to be able to deduct the "purchase tax" again.

One is directly on an incoming invoice (061) and the other one (083) needs to be done separately in an own account move.

Hope this helps, but 100% correct and used in several companies and checked with tax lawyers 😉

Comment (jco-odoo) [reply addons/l10n_at/data/account_tax_template.xml:2029]: Where do you need to report the base amount?  (normally doing a group of taxes for Spain) 
Because if no need for reporting, we could use a division type.

Comment (wtaferner) [reply addons/l10n_at/data/account_tax_template.xml:2029]: Puh, difficult to relate to Spain as I am not used to what they have or have not to do and also I have not a lot of experience with Spanish companies but I can remember we discussed this before and I was convinced that the way I do it is, but feel free to give me a call and we will try to elaborate together once again. You might have a point, but as this is a "complicated" tax use case, I can not just say, yes, you are right, I have totally to rework this now after using it for a long time in production and redefining also several times to simplify for the usage and daily routine of the accountant user in Austria 😉

**Review by JulienVR — COMMENTED** [review]
Few comments on the technical side. You'll also need to rebase to make the `template` ci happy and take a look also at the `style` one. :)

File: addons/l10n_at/data/account_account_tag.xml
Comment (wtaferner) [reply addons/l10n_at/data/account_account_tag.xml:373]: Color? to have a color for the tag? Maybe in later versions again deprecated but in 15.0 I think it is still there 😄

File: addons/l10n_at/data/account_fiscal_position_template.xml
Comment (wtaferner) [reply addons/l10n_at/data/account_fiscal_position_template.xml:27]: Haha, ask @william-andre, so yeah, it is all about automatically detect the fiscal position based on the VAT. If you are from the countries inside the EU and do not have a VAT (which might happen sometimes), we do not want to apply the Non EU FP. Even worse for Austrian countries having a VAT which would now be detected as FP EU instead of National.

Besides these I have really a lot of FP in place for dealing with more complicated stuff, but for Standard I think this should help to have a basic set without running into wrong detections.

File: addons/l10n_at/data/account_tax_report_data.xml
Comment (wtaferner) [reply addons/l10n_at/data/account_tax_report_data.xml:11]: I have fixed that for the other PR in `16.0` IMO already, no idea about the fun in `master` yet, but yeah, be assured, I have to go back and forth unfortunately quite a bit, but well, as a contributor you have to suffer I guess, at least as much as an internal dev 😜

File: addons/l10n_at/data/account_tax_template.xml
Comment (wtaferner) [reply addons/l10n_at/data/account_tax_template.xml:97]: Not sure I can follow, but yeah, probably it has grown over time instead of doing them all at once...but at least, this is quite close to real life work on the product 😉 but yeah, I guess we will have more things being changed (XML -> CSV, refactoring of framework # 100, etc.)

**Review by wtaferner — COMMENTED** [review]
Hmm, I will probably skip the style ones for stable, as it will change anyway in `16.0` and `master`, but appreciate that some kind of style is finally optionally applied.

EDIT: Every change is by the way quite a "big" deal for me as I have to redo our fork commits in our production branches as we, of course, do not wait for standard to merge our patches and after merge I have always to resolve all the conflicts again and again 😉 

So, yeah, we do only propose things which are already actively used for all our clients.

File: addons/l10n_at/data/account_fiscal_position_template.xml
Comment (wtaferner) [reply addons/l10n_at/data/account_fiscal_position_template.xml:27]: Ahh, and besides this, it exists for B2B only, as OSS only applies to B2C and therefore we can not apply the related OSS FP as they would not be allowed to deduct the tax. They get billed (as long as the company does not hit the limit to create a foreign EU VAT) the local tax. This is why there is no tax mapping, but it is a totally different use case and therefore also a different FP.

File: addons/l10n_at/data/account_tax_template.xml
Comment (jco-odoo) [reply addons/l10n_at/data/account_tax_template.xml:2029]: You are not talking about the base amount, so I suppose we can just put this tax as a division included one, so there is no need for an extra negative line and the line becomes just tax.

File: addons/l10n_at/data/account_account_tag.xml
Comment (JulienVR) [reply addons/l10n_at/data/account_account_tag.xml:373]: Yes, that was my guess too but looking at the code, I couldn't find any use of this field. In addition, I haven't seen any colors applied to the tags either :thinking: :

File: addons/l10n_at/data/account_fiscal_position_template.xml
Comment (JulienVR) [reply addons/l10n_at/data/account_fiscal_position_template.xml:27]: Oooh ok, thanks for the explanation :)

File: addons/l10n_at/data/account_account_tag.xml
Comment (wtaferner) [reply addons/l10n_at/data/account_account_tag.xml:373]: You mean this one? :-) It seems to be still there in 15.0...but yeah, probably I removed it in 16.0...so let's keep it for now.

https://github.com/odoo/odoo/blob/bb3742fcc82c032d85f6a547ba1ecb8a84819a25/addons/account/models/account_account_tag.py#L12

File: addons/l10n_at/data/account_tax_template.xml
Comment (wtaferner) [reply addons/l10n_at/data/account_tax_template.xml:2029]: @jco-odoo 
IMO, I need the taxes in different `account.move.line` than the line with the 100% base amount.

image

Here is what I expect from 083 and 061 whereas the latter can be created from an purchase invoice line.

Would your suggestion create the exact same `account.move`?

Comment (jco-odoo) [reply addons/l10n_at/data/account_tax_template.xml:2029]: image

Comment (wtaferner) [reply addons/l10n_at/data/account_tax_template.xml:2029]: Not the same IMO and having a `account.move.line` with 0 debit and 0 credit is also awkward.

Comment (wtaferner) [reply addons/l10n_at/data/account_tax_template.xml:2029]: Maybe it helps that in `083` the deductible VAT must be claimed via the report whereas the applied tax is claimed on customs import directly...so, in the end `3530` will be reconciled / reduced with the amount being paid or getting back from the tax authority.
As said, it is "complicated", not sure how we can settle this here on Github w/o you believing me that I need it as it is 😄
Still open to try to explain it to you in person again, but maybe just accepting my solution would make me happier I guess.

**Comment by wtaferner** [comment]
@JulienVR 
Style is fixed, template seems unrelated to me...

**Comment by JulienVR** [comment]
@wtaferner, rebase both your branches and it should work ;)

**Comment by wtaferner** [comment]
@JulienVR @jco-odoo @Levizar 
Ready? if so, I would move on to the 16.0 branch PR...

**Review by JulienVR — APPROVED** [review]
Ok for me :) @jco-odoo or @william-andre will review and r+ :)

File: addons/l10n_at/data/account_tax_template.xml
Comment (wtaferner) [reply addons/l10n_at/data/account_tax_template.xml:2029]: The product line does represent 100% of the tax as you will never find a base amount in terms of purchase. This tax is mostly used on carrier invoices who are forwarding the tax for us to the tax authority which is based on the value of the goods imported and bill us for exactly this amount to get their money back. The value of the goods will not be part of the invoice lines.

The 20% is the VAT applied on the value of the goods (including insurance, transport costs and product costs), so our 100% will represent 20% of the "invisible" amount (not represented on the invoice in accounting), maybe as a side note somewhere in the carrier invoice depending on the carriers IT.

I hope this helps to understand.

File: addons/l10n_at/data/account_tax_template.xml
Comment (jco-odoo) [inline addons/l10n_at/data/account_tax_template.xml:2073]: @wtaferner What is the point in +100 +100 -100 -100 if you can just do 0 +100?

Comment (jco-odoo) [reply addons/l10n_at/data/account_tax_template.xml:2073]: Maybe because 100 != 0...  So this is reporting only?

Comment (wtaferner) [reply addons/l10n_at/data/account_tax_template.xml:2073]: It is about having those move lines properly next to each other...it is basically two account moves with debit/credit pairs combined into one move (basically how accounting is normally done in its simple form...one at a time without split or combination)...for a reason we do not want to cancel out the amount on the same account and have a zero account move line left. Those two lines reconcile which each other instead of not exist at all as it is an awkward zero line.

It is more about auditing than reporting, but is also important for this specific case in reporting.

This is by the way some feature request I do have as it is requested from many accountants and tax lawyers...

Namely, having the direct opposite accounts for each debit/credit line (`many2many`) to know how it evolved to the final debit/credit which is missing in Odoo to look at an isolated line and know what was going on, even though you can retrieve somehow the all the other lines based on the move name, but sometimes the other move lines of a move can be completely unrelated, depending how you grouped your move lines into a move.

Comment (jco-odoo) [reply addons/l10n_at/data/account_tax_template.xml:2073]: @wtaferner It would be good to understand at least why this tax is that special.  Possibly this solution does the trick, but it is still weird.

Comment (wtaferner) [reply addons/l10n_at/data/account_tax_template.xml:2073]: @jco-odoo 
How? I tried to explain it several times and tried it to configure or make it possible to the accountants in several ways and this was the best way how they properly encode and understand this tax in Odoo and also keep track of what the use case is asking for.

If you have several hours to deep dive in person, I could try to show you all the stuff I have and my customers have to handle, but honestly, I do not believe it is worth the effort. Whenever there is a better way to do it, believe me, I will do it and suggest, especially as I am constantly thriving to improve stuff and get feedback from accountants and tax lawyers in Austria to push the localization and Odoo forward. This is the Austrian localization, I am pretty sure there are other weird countries and use cases out there 😜

Comment (wtaferner) [reply addons/l10n_at/data/account_tax_template.xml:2073]: **\*\*redacted rant\*\***

@jco-odoo @tsb-odoo 
I am here to help, you might want to start here, but do not expect me to understand that I am wrong about something which has been elaborated over several years...

**KZ 061**
Vorsteuern betreffend die entrichtete Einfuhrumsatzsteuer (§ 12 Abs. 1 Z 2 lit. a)
**KZ 083**
Vorsteuern betreffend die geschuldete, auf dem Abgabenkonto verbuchte Einfuhrumsatzsteuer (§ 12 Abs. 1 Z 2 lit. b)

https://www.ris.bka.gv.at/NormDokument.wxe?Abfrage=Bundesnormen&Gesetzesnummer=10004873&Paragraf=12

https://findok.bmf.gv.at/findok?execution=e4s1#RZ_1874a
https://formulare.bmf.gv.at/service/formulare/inter-Steuern/pdfd/2023/U30a.pdf

**Comment by tsb-odoo** [comment]
@wtaferner After reading multiple times the PR, I must confess I still don't get the answer to the questions raised by the different people. 

For "VST_061 entrichtete EUst (§ 12 Abs. 1 Z 2 lit. a)", I doubt this is really something we want. 
Snag_a2b87e
Snag_a2a19b

Why consider twice the VAT deductible? 
Why consider the VAT is due since it has been done earlier by the carrier in some cases, hasn't it ? 

While we are at it, "VST_083 verbuchte EUst. (§ 12 Abs. 1 Z 2 lit. b)" is not used as a sales nor as a purchase tax. Can you please via screenshots tell us the business story behind it ?  

Thank you for favoring useful answers to trials of intentions.

**Comment by wtaferner** [comment]
@tsb-odoo 
I stop explaining anything here on Github as written via email to Josse already. Please have a call/meet or we delete the whole tax as it is non-functional at the moment one way or the other. Be assured that asking over and over the same questions will not help you to understand.

But honestly, does any complicated tax will just work properly if you have no knowledge how to use it?
Anyway, please let's have a meeting or drop the tax...I can keep it in my own repo for my customers and we keep the reference for not being dropped during migrations.

**Comment by tsb-odoo** [comment]
so be it, let's remove what can't seem to be explained calmly

**Comment by wtaferner** [comment]
@tsb-odoo 
Ok, better to have my peace than trying to convince/explain to those who do not want to do an effort to understand.

File: addons/l10n_at/data/account_fiscal_position_template.xml
Comment (jco-odoo) [reply addons/l10n_at/data/account_fiscal_position_template.xml:6]: @william-andre @wtaferner I do not agree.  We should keep the National position as it was as if it is national, it does not matter if it has VAT or not.  It only matters for the EU.

Comment (wtaferner) [reply addons/l10n_at/data/account_fiscal_position_template.xml:6]: Auto detection does not work for Austrian companies having a VAT, they will become the wrong fiscal position...

Comment (jco-odoo) [reply addons/l10n_at/data/account_fiscal_position_template.xml:6]: @wtaferner Testing it out, it is indeed wrong the way it is now.  But if we have the vat_required mit UID on the National, then it will go to the other new EU fiscal position you added without VAT required (which we mostly call EU private).

Comment (tsb-odoo) [reply addons/l10n_at/data/account_fiscal_position_template.xml:6]: Indeed, the solution provided by @jco-odoo is in line with the way fiscal positions work best, and consistent between localizations. Easier to maintain and explain when consistent between similar localizations (all EU countries are similar in this regard).

Comment (wtaferner) [reply addons/l10n_at/data/account_fiscal_position_template.xml:6]: @jco-odoo 
It is about the sequence and IMO it is correctly aligned

National (ohne UID) -> limited to Austria (catch all for Austria w/o UID)
National (mit UID) -> limited to Austria (catch all for Austria w/ UID)
Europäische Union -> limited to "Europe" will catch all in the Country Group with UID
Europäische Union (ohne UID) -> limited to "Europe" will catch all those of the Country Group w/o UID (otherwise those would land in the last FP
Drittstaaten -> catch all others

We could change position of EU (w/ and w/o) but auto detection should be ok as suggested.

EU private -> it is rarely private as OSS is in place, mostly it is business not yet having a VAT number or a lack of providing it...based on this you additionally need to take care about the limits as before.

If a business tries to deduct the VAT billed through OSS it is forbidden to deduct the local VAT, this is why we need this extra fiscal position within the EU which is basically for B2B the local tax.

Comment (tsb-odoo) [reply addons/l10n_at/data/account_fiscal_position_template.xml:6]: assumption that fiscal positions are based on their sequence only is incorrect

Comment (pedrobaeza) [reply addons/l10n_at/data/account_fiscal_position_template.xml:6]: AFAIK, the FP auto-detection is a bunch code with a lot of `if` adding things to a search domain:

https://github.com/odoo/odoo/blob/5d82f678d33d20aa42ba9739a6d14cb6554be0b7/addons/account/models/partner.py#L168

I have already talked several times with Odoo accounting devs about this in Odoo XPs, but it seems it continues the same.

IMO this should be sequence-based for being able to map several conditions (with or without VAT, VIES or not, etc) to finally got the desired fiscal position, specially when you business is both B2B/B2C.

Comment (jco-odoo) [reply addons/l10n_at/data/account_fiscal_position_template.xml:6]: @pedrobaeza Should we really map a tax different if VAT or VIES?  The VIES check was added for the vat_required.

Comment (pedrobaeza) [reply addons/l10n_at/data/account_fiscal_position_template.xml:6]: Of course a possible refactoring of the fiscal position search is not something to target on a stable version. Anyway, I just want to thank you for at least having this debate around it. I hope maybe something good comes from it for future versions.

I don't think it's so risky as you say to let the users to handle this. They are doing a similar thing in pricelists for example. Maybe the sequence field (and widget) can be under a security group option "Advance fiscal position mapping" or so.

The problems with current scheme are:

- It doesn't handle all possible situations correctly. Not sure if it has been corrected lately, but an scenario of mixed B2B/B2C e-commerce with European OSS was not mapping correctly the taxes in Spain: 
  - Intracommunity taxes for B2B VIES companies.
  - Regular taxes for B2B not VIES companies.
  - OSS taxes for B2C customers.

  And the blame history of the method can show you how broken it is due to the constant fixes, patches, reverts: https://github.com/odoo/odoo/blame/5d82f678d33d20aa42ba9739a6d14cb6554be0b7/addons/account/models/partner.py#L168

- It doesn't allow easily to add more custom criteria. Look the ugly patches for getting some of this:
  - https://github.com/OCA/account-fiscal-rule/blob/16.0/account_fiscal_position_partner_type/models/account_fiscal_position.py
  - https://github.com/OCA/account-fiscal-rule/blob/13.0/account_fiscal_position_autodetect_optional_vies/models/account_fiscal_position.py

Comment (jco-odoo) [reply addons/l10n_at/data/account_fiscal_position_template.xml:6]: @wtaferner If you can just remove this fiscal position (National (ohne UID)) and squash the commits, we can merge.

Comment (wtaferner) [reply addons/l10n_at/data/account_fiscal_position_template.xml:6]: @jco-odoo 
No idea why you think this is correct as auto detection will not work for an Austrian company having a VAT (I gave evidence for the need above) -> they will get the wrong fiscal position, but why not, the users on SaaS and the rest will struggle, not me...

**Comment by wtaferner** [comment]
@jco-odoo @william-andre 
Thank you, let's hope this is now merged and we can move on, I will adapt the 16.0 PR as agreed here that we can merge that one too that @svfu-odoo and I can complete the one on master for the forward porting for the SaaS versions I guess you will need to take a look at 16.0 at least as I do not know which feature was introduced when, but I guess it is/was organized that it won't be too hard.

**Comment by jco-odoo** [comment]
@robodoo r+ rebase-ff

**Pull request closed without merging**
